packages/tpetra/core/src/Tpetra_Details_lclMultiply.hpp

```diff
@@ -0,0 +1,149 @@
+// @HEADER
+// ***********************************************************************
+//
+//          Tpetra: Templated Linear Algebra Services Package
+//                 Copyright (2008) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef TPETRA_DETAILS_LCLMULTIPLY_HPP
+#define TPETRA_DETAILS_LCLMULTIPLY_HPP
+
+/// \file Tpetra_Details_lclMultiply.hpp
+/// \brief Declaration and definition of Tpetra::Details::lclMultiply, an
+///   implementation detail of Tpetra::MultiVector.
+///
+/// \warning This file, and its contents, are implementation details
+///   of Tpetra, and may change or disappear at any time.
+
+#include "Kokkos_ArithTraits.hpp"
+#include "KokkosBlas.hpp"
+#include "Tpetra_Details_Profiling.hpp"
+#include "Kokkos_InnerProductSpaceTraits.hpp"
+
+namespace Tpetra {
+namespace Details {
+
+template<class XMV,
```

Comment:
It's not clear to me why you need the `enable_if`, since both versions appear to invoke the same version of `KokkosBlas::gemm`.

Reply:
Maybe I mocked things up, cause I don't understand `enable_if` very well. What I want is that if `value_type==dot_type`, we enable the specialized `dot` for the case the one of the multivectors is actually a vector. Otherwise we always use `gemm`. So the `gemm` call is identical in both cases.

Reply:
@cgcgcg I took another look at Stokhos, and I think it's OK not to worry about the `dot_type != value_type` case.  Stokhos has specializations of `Belos::MultiVecTraits` that just extract out the MultiVectors' local data, flatten the Views into `dot_type`, and do the local multiply on the "flat views."

Reply:
Oh wait, I think I see what you mean.  Why not just have special cases that call `KokkosBlas::gemv`?  That would avoid the whole `dot_type != value_type` issue.

Reply:
@mhoemmen Let me give try that out..